File: src/main/java/com/timepath/hl2/io/bsp/lump/FaceHandler.java
package com.timepath.hl2.io.bsp.lump;

import com.timepath.hl2.io.bsp.Lump;
import com.timepath.hl2.io.bsp.LumpHandler;
import com.timepath.io.OrderedInputStream;
import com.timepath.io.struct.Struct;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

class FaceHandler implements LumpHandler<Face[]> {

    private static final Logger LOG = Logger.getLogger(FaceHandler.class.getName());
    private static final int MAX_MAP_FACES = 65536;

    FaceHandler() {
    }

    @NotNull
    @Override
    public Face[] handle(@NotNull Lump l, @NotNull OrderedInputStream in) throws IOException {
        try {
            @NotNull Face[] e = new Face[l.length / Struct.sizeof(new Face())];
            for (int i = 0; i < e.length; i++) {
                e[i] = in.readStruct(new Face());
            }
            return e;
        } catch (@NotNull InstantiationException | IllegalAccessException ex) {
            LOG.log(Level.SEVERE, null, ex);
        }
        return new Face[0];
    }
}

File: src/main/kotlin/com/timepath/hl2/io/bsp/lump/FaceHandler.kt
package com.timepath.hl2.io.bsp.lump

import com.timepath.hl2.io.bsp.Lump
import com.timepath.hl2.io.bsp.LumpHandler
import com.timepath.io.OrderedInputStream
import com.timepath.io.struct.Struct

import java.io.IOException
import java.util.logging.Level
import java.util.logging.Logger

class FaceHandler : LumpHandler<List<Face>> {

    throws(IOException::class)
    override fun handle(l: Lump, `in`: OrderedInputStream): List<Face> {
        try {
            return (0..(l.length / Struct.sizeof(Face())) - 1).map {
                `in`.readStruct<Face>(Face())
            }
        } catch (ex: InstantiationException) {
            LOG.log(Level.SEVERE, null, ex)
        } catch (ex: IllegalAccessException) {
            LOG.log(Level.SEVERE, null, ex)
        }

        return emptyList()
    }

    companion object {

        private val LOG = Logger.getLogger(javaClass<FaceHandler>().getName())
        private val MAX_MAP_FACES = 65536
    }
}
